com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/StepRequestHandler.java

```diff
@@ -56,63 +67,109 @@ public StepRequestHandler() {
         }
 
         if (command == Command.CONFIGURATIONDONE) {
-            context.getDebugSession().getEventHub().stepEvents().subscribe(debugEvent -> {
-                handleDebugEvent(debugEvent, context.getDebugSession(), context);
-            });
-        } else {
-            long threadId = ((StepArguments) arguments).threadId;
-            ThreadReference thread = DebugUtility.getThread(context.getDebugSession(), threadId);
-            if (thread != null) {
-                setPendingStepKind(thread, command);
-                setOriginalStackDepth(thread);
-                setOriginalStepLocation(thread);
+            context.getDebugSession().getEventHub().events()
+                .filter(debugEvent -> debugEvent.event instanceof StepEvent
+                        || debugEvent.event instanceof BreakpointEvent
+                        || debugEvent.event instanceof ThreadDeathEvent)
+                .subscribe(debugEvent -> {
+                    handleDebugEvent(debugEvent, context.getDebugSession(), context);
+                });
+            return CompletableFuture.completedFuture(response);
+        }
+
+        long threadId = ((StepArguments) arguments).threadId;
+        ThreadReference thread = DebugUtility.getThread(context.getDebugSession(), threadId);
+        if (thread != null) {
+            try {
+                setPendingStepKind(threadId, command);
```

Comment:
Do we need to check the thread status before setting it? What if the thread state is already set by other requests?